gk/main.c

```diff
@@ -1483,232 +1776,92 @@ process_pkts_front(uint16_t port_front, uint16_t port_back,
-					if (ret < 0) {
-						drop_packet_front(
-							pkt, instance);
-					}
-					continue;
-				} else if (ret < 0) {
-					drop_packet_front(pkt, instance);
-					continue;
-				}
-
-				fe = &instance->ip_flow_entry_table[ret];
-				initialize_flow_entry(fe, &packet.flow, fib);
-				break;
-
-			case GK_FWD_GATEWAY_BACK_NET: {
-			 	/*
-				 * The entry instructs to forward
-				 * its packets to the gateway in
-				 * the back network, forward accordingly.
-				 *
-				 * BP block bypasses from the front to the
-				 * back interface are expected to bypass
-				 * ranges of IP addresses that should not
-				 * go through Gatekeeper.
-				 *
-				 * Notice that one needs to update
-				 * the Ethernet header.
-				 */
-
-				eth_cache = fib->u.gateway.eth_cache;
-				RTE_VERIFY(eth_cache != NULL);
-
-				if (adjust_pkt_len(pkt, back, 0) == NULL ||
-						pkt_copy_cached_eth_header(pkt,
-							eth_cache,
-							back->l2_len_out)) {
-					drop_packet_front(pkt, instance);
-					continue;
-				}
-
-				if (update_ip_hop_count(front, &packet,
-						num_pkts, icmp_bufs,
-						&instance->front_icmp_rs,
-						instance,
-						drop_packet_front) < 0)
-					continue;
-
-				tx_bufs[num_tx++] = pkt;
-				continue;
-			}
+	/* Lookup the flow entries in batches. */
+	for (i = 0; i < q; i += batch_size) {
+		ret = rte_hash_lookup_bulk(instance->ip_flow_hash_table,
+			(const void **)&flow_arr[i], batch_size, &pos_arr[i]);
+		if (ret != 0) {
+			GK_LOG(NOTICE,
+				"failed to find multiple keys in the hash table at lcore %u\n",
+				rte_lcore_id());
+		}
+	}
 
-			case GK_FWD_NEIGHBOR_BACK_NET: {
-				/*
-				 * The entry instructs to forward
-				 * its packets to the neighbor in
-				 * the back network, forward accordingly.
-				 */
-				if (packet.flow.proto == RTE_ETHER_TYPE_IPV4) {
-					eth_cache = lookup_ether_cache(
-						&fib->u.neigh,
-						&packet.flow.f.v4.dst);
-				} else {
-					eth_cache = lookup_ether_cache(
-						&fib->u.neigh6,
-						&packet.flow.f.v6.dst);
-				}
-
-				RTE_VERIFY(eth_cache != NULL);
-
-				if (adjust_pkt_len(pkt, back, 0) == NULL ||
-						pkt_copy_cached_eth_header(pkt,
-							eth_cache,
-							back->l2_len_out)) {
-					drop_packet_front(pkt, instance);
-					continue;
-				}
-
-				if (update_ip_hop_count(front, &packet,
-						num_pkts, icmp_bufs,
-						&instance->front_icmp_rs,
-						instance,
-						drop_packet_front) < 0)
-					continue;
-
-				tx_bufs[num_tx++] = pkt;
-				continue;
-			}
+	if (r != 0) {
+		ret = rte_hash_lookup_bulk(instance->ip_flow_hash_table,
+			(const void **)&flow_arr[q], r, &pos_arr[q]);
+		if (ret != 0) {
+			GK_LOG(NOTICE,
+				"failed to find multiple keys in the hash table at lcore %u\n",
+				rte_lcore_id());
+		}
+	}
 
-			case GK_DROP:
-				/* FALLTHROUGH */
-			default:
-				drop_packet_front(pkt, instance);
-				continue;
+	for (i = 0; i < num_ip_flows; i++) {
+		if (pos_arr[i] >= 0)
+			fe_arr[i] = &instance->ip_flow_entry_table[pos_arr[i]];
```

Comment:
Prefetch:
```
RTE_BUILD_BUG_ON(sizeof(*fe_arr[0]) > 2*RTE_CACHE_LINE_SIZE);
rte_prefetch0(fe_arr[i]);
rte_prefetch0(((char *)fe_arr[i]) + RTE_CACHE_LINE_SIZE);
```
Since we need to do the same prefetch later in the code a couple of more times, put it into an inline function `prefetch_flow_entry()`.

Reply:
Add the test `RTE_BUILD_BUG_ON(sizeof(*fe_arr[0]) <= RTE_CACHE_LINE_SIZE);` at the very beginning of `prefetch_flow_entry()` to ensure that we need two prefetches.